### PR TITLE
Misc refactoring

### PR DESCRIPTION
**Pull request by boxdot**
* Improve error formatting
* Add various number warnings
* Do not compare precise output of informative tests
* Output during program execution
* Validate line numbers before execution

### CODE REVIEW

**Comment by boxdot** [comment]
Resolves #22.

File: src/ast.rs [hunk 1]
```diff
@@ -100,6 +102,47 @@ impl<'a> Program<'a> {
             data,
         })
     }
+
+    fn validate(&self, source_code: &str) -> Result<(), Error> {
```
Comment (jeschkies) [inline src/ast.rs]: A quick documentation comment would be good here.

**Comment by boxdot** [comment]
Btw. we should rather `rebase and merge` since each commit is kind of independent of each other.

**Comment by jeschkies** [comment]
> Btw. we should rather rebase and merge since each commit is kind of independent of each other.

But now squashing, right? Cause that would be a painful endeavor again.

**Pull request merged**